compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -254,11 +250,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         mut expected_ty_expr: Option<&'tcx hir::Expr<'tcx>>,
         allow_two_phase: AllowTwoPhase,
-    ) -> (Ty<'tcx>, Option<Diag<'tcx>>) {
+    ) -> Result<Ty<'tcx>, Diag<'tcx>> {
         let expected = self.resolve_vars_with_obligations(expected);
 
         let e = match self.coerce(expr, checked_ty, expected, allow_two_phase, None) {
-            Ok(ty) => return (ty, None),
+            Ok(ty) => return Ok(ty),
```

Comment:
we `set_tainted_by_errors` in line 263 using a `span_delayed_bug` and then emit the error further down in line 270. Feel like we should just flip the order and use the `ErrorGuaranteed` from actually emitting something

Reply:
Oh, we didn't emit the error initially, so I didn't change it, good catch

Reply:
Ah nevermind, we're not actually emitting the error here, we're just creating the diagnostic